WebService/v1/Models/DeviceModelApiModel/DeviceModelApiModel.cs

```diff
@@ -82,12 +139,61 @@ public static DeviceModelApiModel FromServiceModel(DeviceModel value)
                 result.Telemetry.Add(DeviceModelTelemetry.FromServiceModel(message));
             }
 
-            foreach (var method in value.CloudToDeviceMethods)
+            if (value.CloudToDeviceMethods?.Count > 0)
             {
-                result.CloudToDeviceMethods.Add(method.Key, DeviceModelSimulationScript.FromServiceModel(method.Value));
+                foreach (var method in value.CloudToDeviceMethods)
+                {
+                    result.CloudToDeviceMethods.Add(method.Key, DeviceModelSimulationScript.FromServiceModel(method.Value));
+                }
             }
 
             return result;
         }
+
+        public void ValidateInputRequest(ILogger log, bool isUpserting = false)
+        {
+            const string NO_ETAG = "The device model doesn't contain an ETag";
+            const string NO_PROTOCOL = "The device model doesn't contain a protocol";
+            const string ZERO_TELEMETRY = "The device model has zero telemetry";
+
+            // A custom device model must contain a ETag when upserting
+            if (isUpserting && this.ETag == String.Empty)
+            {
+                log.Error(NO_ETAG, () => new { deviceModel = this });
+                throw new BadRequestException(NO_ETAG);
+            }
```

Comment:
lines 160-164 can be removed, all the ETag logic is already in the service layer

Reply:
Done.